File: ports/1.21.1-full/src/main/java/mods/eln/server/SaveConfig.java
package mods.eln.server;

import mods.eln.Eln;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.saveddata.SavedData;

public class SaveConfig extends SavedData {

    public static SaveConfig instance;

    public boolean heatFurnaceFuel = true;
    public boolean electricalLampAging = true;
    public boolean batteryAging = true;
    public boolean infinitePortableBattery = false;
    public boolean reGenOre = false;
    public double cableRsFactor_lastUsed = 1.0;

    public SaveConfig(String par1Str) {
        super(par1Str);
        instance = this;
    }

    @Override
    public void readFromNBT(CompoundTag nbt) {
        heatFurnaceFuel = nbt.getBoolean("heatFurnaceFuel");
        electricalLampAging = nbt.getBoolean("electricalLampAging");
        batteryAging = nbt.getBoolean("batteryAging");
        infinitePortableBattery = nbt.getBoolean("infinitPortableBattery");
        reGenOre = nbt.getBoolean("reGenOre");
        cableRsFactor_lastUsed = nbt.getDouble("cableRsFactor_lastUsed");
        Eln.windProcess.readFromNBT(nbt, "wind");
    }

    @Override
        public CompoundTag writeToNBT(CompoundTag nbt) {
        nbt.putBoolean("heatFurnaceFuel", heatFurnaceFuel);
        nbt.putBoolean("electricalLampAging", electricalLampAging);
        nbt.putBoolean("batteryAging", batteryAging);
        nbt.putBoolean("infinitPortableBattery", infinitePortableBattery);
        nbt.putBoolean("reGenOre", reGenOre);
        Eln.windProcess.writeToNBT(nbt, "wind");
        return nbt;
    }

    @Override
    public boolean isDirty() {
        return true;
    }
}

File: src/main/kotlin/mods/eln/server/SaveConfig.kt
package mods.eln.server

import mods.eln.Eln
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.saveddata.SavedData

/** The mod's per-world settings (the wind state), as world-saved data. */
class SaveConfig : SavedData() {
    override fun save(nbt: CompoundTag, registries: HolderLookup.Provider): CompoundTag {
        Eln.wind.writeToNBT(nbt, "wind")
        return nbt
    }

    override fun isDirty(): Boolean {
        return true
    }

    companion object {
        @JvmField
        var instance: SaveConfig? = null

        @JvmField
        val FACTORY = Factory({ SaveConfig() }, { nbt, _ -> SaveConfig().also { Eln.wind.readFromNBT(nbt, "wind") } }, null)
    }

    init {
        instance = this
    }
}
